### PR TITLE
Exit longlived greenlets if any of them fails

### PR DESCRIPTION
**Pull request by andrevmatos**
Implements 3rd point of #2201

### CODE REVIEW

File: raiden/network/transport/matrix.py [hunk 1]
```diff
@@ -280,6 +280,11 @@ def stop_and_wait(self):
             self._client.greenlets.clear()
             self._client.logout()
 
+    @property
+    def greenlet(self) -> gevent.Greenlet:
```
Comment (hackaugusto) [inline raiden/network/transport/matrix.py]: I think this is a bad approach, why don't you make the `RaidenService` and `Transport` classes  greenlets, and let them manage their child tasks instead of exposing it?

Comment (andrevmatos) [reply raiden/network/transport/matrix.py]: @hackaugusto I thought about it too, but these classes aren't greenlets on theirselves, as there's no obvious loop inside (`MatrixTransport` depends on the underlying `_client.sync_thread` greenlet, `RaidenService` doesn't actually have a loop/long-lasting `_run` method or alike, `APIServer` depends on the `WSGIServer` instance, which isn't even a greenlet on itself nor have one, but use low-level gevent calls to asynchronously dispatch request handling, that greenlet inside it just wraps the busy wait).
Rather, they _control_ one or more subtasks/greenlets, like `AlarmTask`, `GMatrixClient` and `WSGIServer`, and are the responsible for starting/stopping them and being a point of control if the underlying task fails unexpectedly. How about a `RaidenTask` abstract class, with `start`, `stop`, possibly `join` methods and a `long_greenlets() -> List[Greenlet]` property?

Comment (andrevmatos) [reply raiden/network/transport/matrix.py]: An option would also to make a small `StoppableGreenlet` abstract class (greenlet with synchronous `stop` method), and make those classes inherint from it, making them effectivelly greenlets (as you asked), but in those cases, the `_run` method would be a simple `get` on the children tasks, which would make it re-raise underlying exceptions, if any, and cleanup and exit cleanly when `stop` is called

Comment (hackaugusto) [reply raiden/network/transport/matrix.py]: Yeah, having a standard method to stop all these classes is good, I just want to avoid the introspection used here:

https://github.com/raiden-network/raiden/blob/d8839aee71d85e2d9c0617626001c36b17f82fc4/raiden/ui/cli.py#L1015-L1025

File: raiden/ui/cli.py [hunk 2]
```diff
@@ -1024,8 +1034,19 @@ def _run_app(self):
...
-                api_server.stop()
+
+            def stop_task(task, timeout=None):
+                if isinstance(task, Runnable):
+                    task.stop(timeout=timeout)
+                else:
+                    task.kill(timeout=timeout)
+                task.get()  # re-raise
+
+            stop_tasks = [
```
Comment (hackaugusto) [inline raiden/ui/cli.py]: what about changing from an event to an async result?

then the try above could be just use a `event.get()`, and the exception would be re-raised for us automatically

File: raiden/tests/utils/network.py [hunk 3]
```diff
@@ -313,7 +313,7 @@ def create_apps(
                                 'server': 'matrix.local.raiden',
                             },
                             'retries_before_backoff': retries_before_backoff,
-                            'retry_interval': retry_interval,
+                            'retry_interval': retry_interval * 5,
```
Comment (hackaugusto) [inline raiden/tests/utils/network.py]: this is not the right place for this change, please change it on `raiden.tests.fixtures.variables`

File: raiden/tasks.py [hunk 4]
```diff
@@ -179,4 +179,8 @@ def run_callbacks(self, current_block, chain_id):
         self.last_block_number = current_block
 
     def stop_async(self):
-        self.stop_event.set(True)
+        self._stop_event.set(True)
+
+    def stop(self, timeout=None):
```
Comment (hackaugusto) [inline raiden/tasks.py]: please don't add timeouts for the inner layers, only use it at the outer-most layers of the code

File: raiden/raiden_service.py [hunk 5]
```diff
@@ -335,10 +337,12 @@ def stop(self):
...
         # contact the disconnected client
         try:
-            with gevent.Timeout(self.shutdown_timeout):
+            with gevent.Timeout(timeout, 'Substasks didn\'t stopped on time'):
+                self.transport.stop()
+                self.alarm.stop()
                 self.blockchain_events.uninstall_all_event_listeners()
         except (gevent.timeout.Timeout, RaidenShuttingDown):
-            pass
+            log.exception('Exception on RaidenService stop')
```
Comment (hackaugusto) [inline raiden/raiden_service.py]: `RaidenShuttingDown`exceptions are expcted durring shutting down and should not be logged.

File: raiden/network/rpc/client.py [hunk 6]
```diff
@@ -237,9 +229,6 @@ def block_number(self):
         """ Return the most recent block. """
         return self.web3.eth.blockNumber
 
-    def inject_stop_event(self, event):
```
Comment (hackaugusto) [inline raiden/network/rpc/client.py]: Why was this removed?

File: raiden/network/transport/udp/udp_transport.py [hunk 7]
```diff
@@ -158,7 +160,8 @@ def __init__(self, discovery, udpsocket, throttle_policy, config):
         self.nat_keepalive_timeout = config['nat_keepalive_timeout']
         self.nat_invitation_timeout = config['nat_invitation_timeout']
 
-        self.event_stop = Event()
+        self.stop_event = Event()
+        self.stop_event.set()
```
Comment (hackaugusto) [inline raiden/network/transport/udp/udp_transport.py]: I believe this event should not be set here

File: raiden/network/transport/udp/udp_transport.py [hunk 8]
```diff
@@ -204,15 +210,27 @@ def start(
...
+    def run(self):
+        """ Runnable main method, perform wait on long-running subtasks """
+        try:
+            self.stop_event.wait()
+        finally:
+            self.stop()  # noop on double call, ensure cleanup and wait on subtasks
+
+    def stop(self, timeout=None):
+        if self.stop_event.ready():
+            return  # double call, happens on normal stop, ignore
```
Comment (hackaugusto) [inline raiden/network/transport/udp/udp_transport.py]: why is `stop` called twice?

File: raiden/network/transport/matrix.py [hunk 9]
```diff
@@ -698,7 +806,7 @@ def _get_unlisted_room(self, room_name, invitees):
                 else:
                     self.log.info('Room joined successfully', room=room)
                     return room
-            gevent.sleep(self._login_retry_wait)
+            gevent.sleep(self._config.get('login_retry_wait', 0.5))
```
Comment (hackaugusto) [inline raiden/network/transport/matrix.py]: `0.5` should be a constant define in `settings.py`

File: raiden/network/transport/matrix.py [hunk 10]
```diff
@@ -573,28 +644,65 @@ def _send_queued_messages(
...
+                        message=message,
+                        queue=queue_identifier,
+                        queueids_to_queues=self._queueids_to_queues,
+                    )
+                    break
+                # retry while the message is in queue
+                # Delivered and Processed messages should eventually remove them
+                if not any(
+                    message_id == event.message_identifier
+                    for event in self._queueids_to_queues[queue_identifier]
```
Comment (hackaugusto) [inline raiden/network/transport/matrix.py]: could you move this up so that we don't have a multi line if-statement?

File: raiden/network/transport/matrix.py [hunk 11]
```diff
@@ -573,28 +644,65 @@ def _send_queued_messages(
...
                 self._config['retry_interval'],
                 self._config['retry_interval'] * 10,
             )
-            while async_result.value is None:
-                self._send_immediate(receiver_address, data)
-                gevent.sleep(next(timeout_generator))
+            for delay in timeout_generator:
+                state = self._address_to_presence.get(receiver_address)
+                if state in reachable:
+                    self._send_raw(receiver_address, data)
```
Comment (hackaugusto) [inline raiden/network/transport/matrix.py]: This is a blocking call, correct?

The `if self._stop_event.wait(delay):` bellow will not be sufficient to exit the task on a timely manner. Is the kill introduced in raiden-libs used to exit this?

File: raiden/network/transport/matrix.py [hunk 12]
```diff
@@ -502,40 +572,42 @@ def _handle_message(self, room, event):
...
+                self._receive_message(message)
         else:
             self.log.error(
                 'Invalid message',
                 message=data,
             )
 
     def _receive_delivered(self, delivered: Delivered):
-        # FIXME: The signature doesn't seem to be verified - check in UDPTransport as well
+        # FIXME: check if UDPTransport also checks Delivered sender and message presence
```
Comment (hackaugusto) [inline raiden/network/transport/matrix.py]: The sender must be checked, otherwise a malicious node could forge a delivered message, which would clear the queue and stop retries.

This is only necessary because UDP is *not* an encrypted transport, so we need to authenticate using signatures.

File: raiden/network/transport/matrix.py [hunk 13]
```diff
@@ -357,18 +401,35 @@ def _join_discovery_room(self):
...
-                raise
-            # Room doesn't exist
-            if discovery_cfg['server'] != self._server_name:
-                raise RuntimeError(
-                    f"Discovery room {discovery_room_alias_full} not found and can't be "
-                    f"created on a federated homeserver {self._server_name!r}.",
-                )
-            discovery_room = self._client.create_room(self._discovery_room_alias, is_public=True)
+        last_ex = None
+        for _ in range(10):
```
Comment (hackaugusto) [inline raiden/network/transport/matrix.py]: These nested exceptions are hard to read, cant you split this in two loops?

```python
while not room_exists and retry():
  try:
    create_room()
  except: ...

if not room_exists:
  raise

while not joined and retry():
  try:
    join()
  except: ...

if not joined:
  raise
```

File: raiden/network/transport/matrix.py [hunk 14]
```diff
@@ -232,15 +285,19 @@ def send_async(
         queue_identifier: QueueIdentifier,
         message: Message,
     ):
-        if not self._running:
+        if self._stop_event.ready():
             return
+
+        message_id = getattr(message, 'message_identifier')
```
Comment (hackaugusto) [inline raiden/network/transport/matrix.py]: Please use `message.message_identifier`

File: raiden/network/transport/matrix.py [hunk 15]
```diff
@@ -188,26 +189,78 @@ def _http_retry_delay() -> Iterable[float]:
...
+        if self._stop_event.ready():
+            return
+        self._stop_event.set()
+        if timeout is None:
+            timeout = self._config.get('logout_timeout', 10)
+
+        self._client.set_presence_state(UserPresence.OFFLINE.value)
+
+        try:
+            with gevent.Timeout(timeout, 'Logging out despite unjoined greenlets.'):
```
Comment (hackaugusto) [inline raiden/network/transport/matrix.py]: we already have a timeout on the raiden service, no need for this here

File: raiden/network/transport/udp/udp_transport.py (hunk 8 above)
Comment (andrevmatos) [reply raiden/network/transport/udp/udp_transport.py]: 4 lines above: if an exception happens, I want `run` to call `stop` on exit, to clean up tasks, but it'll also be run on normal `stop`, but second one will be noop

File: raiden/network/transport/matrix.py (hunk 15 above)
Comment (andrevmatos) [reply raiden/network/transport/matrix.py]: All `Runnable` have the `timeout` parameter for `stop`, I'm just sticking to the interface/behavior, it was already there, just updated for the new interface.

Comment (hackaugusto) [reply raiden/network/transport/matrix.py]: Please don't add the timeout here, we have had multiple discussions about this and timeouts don't compose.

File: raiden/network/transport/matrix.py (hunk 12 above)
Comment (andrevmatos) [reply raiden/network/transport/matrix.py]: Yes, I know.. the point is that I did that there for `Matrix`, so it's just updating the `FIXME` to remember to do this check also for `UDP` eventually.

File: raiden/network/transport/matrix.py (hunk 11 above)
Comment (andrevmatos) [reply raiden/network/transport/matrix.py]: Yes, `_send_raw` blocks, but should be very quick (only request send time), so why `_stop_event.wait(delay)` isn't enough? The `Timeout` on `stop` waiting for these greenlets should ensure it won't block forever

File: raiden/network/transport/matrix.py (hunk 10 above)
Comment (andrevmatos) [reply raiden/network/transport/matrix.py]: It's not a multi-line if, it's a multi-line `any` (I know, I know, just telling my 'justification'). But ok, will do

File: raiden/network/transport/matrix.py (hunk 9 above)
Comment (andrevmatos) [reply raiden/network/transport/matrix.py]: I just moved this code around, but ok

File: raiden/network/transport/udp/udp_transport.py (hunk 7 above)
Comment (andrevmatos) [reply raiden/network/transport/udp/udp_transport.py]: These events starts `set()`ed to tell whoever may care (this one not, but other classes checks for `stop_event` to see if the service is not running) that the object is initialized as stopped. It's cleared upon `start`, to indicate it's running. `stop_event` is then set again to wake up listening greenlets that the service is stopping.

File: raiden/raiden_service.py (hunk 5 above)
Comment (andrevmatos) [reply raiden/raiden_service.py]: I hope `RaidenShuttingDown` won't be needed anymore after proper `stop` and exception handling here in this PR, but no hard feelings for it too

File: raiden/tasks.py (hunk 4 above)
Comment (andrevmatos) [reply raiden/tasks.py]: It's just to comply with the interface. Inner layers may leave then as `None` as well

File: raiden/tests/utils/network.py (hunk 3 above)
Comment (andrevmatos) [reply raiden/tests/utils/network.py]: Problem is that that fixture doesn't have yet access to the transport. Will look on how to inject that there

File: raiden/ui/cli.py (hunk 2 above)
Comment (andrevmatos) [reply raiden/ui/cli.py]: Will check how the signals play with it

File: raiden/network/rpc/client.py (hunk 6 above)
Comment (andrevmatos) [reply raiden/network/rpc/client.py]: It's in the commit's comment: the correct approach both for exception handling and stop time should remove the need for injecting/raising an exception in web3. Instead, whatever uses it should stop and exit gracefully.

**Comment by LefterisJP** [comment]
You should rebase this on latest master. Lots of conflicting files @andrevmatos

**Comment by andrevmatos** [comment]
Yes, I know, but a lot of fixes which weren't directly related but needed for this PR were split on #2268, I may look into merging that first then this one on top of that. Hope to get that ready soon, only a single test failing in a non locally reproducible and weird way (gas exceeds limit error on teardown?)

File: raiden/api/rest.py [hunk 16]
```diff
@@ -325,7 +326,7 @@ def __init__(self, rest_api, cors_domain_list=None, web_ui=False, eth_rpc_endpoi
         # Otherwise there is no guarantee that the node is in a valid state and
         # that the actions are valid, e.g. deposit in a channel that has closed
         # while the node was offline.
-        if not rest_api.raiden_api.raiden.start_event.is_set():
+        if not rest_api.raiden_api.raiden:
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/api/rest.py]: Yes =)

File: raiden/network/transport/udp/udp_transport.py [hunk 17]
```diff
@@ -204,14 +211,31 @@ def start(
             self.init_queue_for(queue_identifier, encoded_queue)
 
         self.server.start()
+        super().start()
+
+    def run(self):
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/network/transport/udp/udp_transport.py]: `run` method from `Runnable` is called inside a greenlet on `start`

File: raiden/network/transport/udp/udp_transport.py [hunk 18]
```diff
@@ -204,14 +211,31 @@ def start(
...
+            self.stop_event.set()
+            gevent.killall(self.greenlets)  # kill children
+            raise  # re-raise to keep killed status
+        except Exception:
+            self.stop()  # ensure cleanup and wait on subtasks
+            raise
+
+    def stop(self):
+        if self.stop_event.ready():
+            return  # double call, happens on normal stop, ignore
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/network/transport/udp/udp_transport.py]: Yes, this simple return was from a left-over change where `stop` could be called twice (once for graceful exit, other on `run` teardown), but since `run` now only calls `stop` when raising, this shouldn't be the case anymore and can successfuly raise if called on non-started Runnable.

File: raiden/raiden_service.py [hunk 19]
```diff
@@ -298,40 +301,36 @@ def start_async(self) -> Event:
...
-                self.start_event.set()
+            endpoint_registration_greenlet.get()  # re-raise if exception occurred
 
-            endpoint_registration_greenlet.link(set_start_on_registration)
-        else:
-            self.start_event.set()
+        super().start()
 
-        return self.start_event
+    def start_async(self) -> gevent.Greenlet:
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/raiden_service.py]: Right, but there we can still spawn these on greenlets, for parallel start (which can take some seconds on udp due to `endpoint_registration_greenlet`), as a test optimization

File: raiden/raiden_service.py [hunk 20]
```diff
@@ -341,9 +340,13 @@ def stop(self):
         # We need a timeout to prevent an endless loop from trying to
         # contact the disconnected client
         try:
-            with gevent.Timeout(self.shutdown_timeout):
+            with gevent.Timeout(self.shutdown_timeout, 'Substasks didn\'t stopped on time'):
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/raiden_service.py]: Ok, just updated from the previous algorithm, will remove

File: raiden/raiden_service.py [hunk 21]
```diff
@@ -341,9 +340,13 @@ def stop(self):
...
+                self.transport.stop()
+                self.alarm.stop()
+                self.transport.get()
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/raiden_service.py]: `stop` isn't intended to raise run-time exceptions, only stop-time ones (as well as `start` raise startup-time exceptions). This `get` here ensures run-time exceptions are re-raised, if any.

File: raiden/ui/cli.py [hunk 22]
```diff
@@ -983,26 +987,32 @@ def _run_app(self):
...
         # spawn a greenlet to handle the gas reserve check
-        gevent.spawn(check_gas_reserve, app_.raiden)
+        tasks.append(gevent.spawn(check_gas_reserve, app_.raiden))
 
         self._startup_hook()
 
         # wait for interrupt
-        event = Event()
+        event = AsyncResult()
         gevent.signal(signal.SIGQUIT, event.set)
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/ui/cli.py]: Ok, will update for a `lambda`

File: raiden/ui/cli.py [hunk 23]
```diff
@@ -1025,8 +1035,21 @@ def _run_app(self):
...
-                api_server.stop()
+
+            def stop_task(task):
+                try:
+                    if isinstance(task, Runnable):
+                        task.stop()
+                    else:
+                        task.kill()
+                finally:
+                    task.get()  # re-raise
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/ui/cli.py]: That's intended, `GreenletExit` isn't properly an exception, and not re-raising it is fine.

File: raiden/raiden_service.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (hackaugusto) [reply raiden/raiden_service.py]: I don't understand what you mean by runtime/stoptime/cleanup exceptions.

File: raiden/ui/cli.py [hunk 24]
```diff
@@ -1025,8 +1035,21 @@ def _run_app(self):
...
+
+            stop_tasks = [
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/ui/cli.py]: Agree, `App` could become a `Runnable` too, the problem is that there's too much things to organize here (in `cli.py`), startup and stop code should be moved to `app.py`, `check`s could go to its own module, or as functions [optionally] called from `App`, and `cli.py` should only be about `cli` and parameters. I gave up this refactoring, as the PR was already big enough, `cli.py`, `app.py`, `utils/cli.py` definitely need cleanup, we're running low on `*run*app` function/method names

File: raiden/ui/cli.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (hackaugusto) [reply raiden/ui/cli.py]: How is it not a proper exception?

File: raiden/utils/runnable.py [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+    In the future, when proper restart is implemented, may be replaced by actual greenlet
+    """
+    greenlet: Greenlet = None
+    args: Sequence = tuple()  # args for run()
+    kwargs: dict = dict()  # kwargs for run()
+
+    def __init__(self, run=None, *args, **kwargs):
+        if run is not None:
+            self.run = run
+        self.args = tuple(args)
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/utils/runnable.py]: Just to ensure it's a tuple, not really needed, but painless as it's only on `__init__` time

File: raiden/utils/runnable.py [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+        self.kwargs = kwargs
+
+        self.greenlet = Greenlet(self.run, *self.args, **self.kwargs)
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/utils/runnable.py]: That'd prevent `link`s before `start`, that's why it's also initialized here and reason for that `pristine` on `start`

File: raiden/utils/runnable.py [hunk 27]
```diff
@@ -0,0 +1,71 @@
...
+        self.greenlet.name = f'{self.__class__.__name__}|{self.greenlet.name}'
+
+    def start(self):
+        """ Synchronously start task
+
+        Reimplements in children an call super().start() at end to start run()
+        Start-time exceptions may be raised
+        """
+        if self.greenlet:
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/utils/runnable.py]: [Greenlet.start is deprecated](https://github.com/gevent/gevent/blob/1afdb093edb855f2249f3130966988dadff965fd/src/gevent/greenlet.py#L397-L400)

File: raiden/utils/runnable.py [hunk 28]
```diff
@@ -0,0 +1,71 @@
...
+            not self.greenlet.dead and
+            tuple(self.greenlet.args) == tuple(self.args) and
+            self.greenlet.kwargs == self.kwargs
+        )
+        if not pristine:
+            self.greenlet = Greenlet(self.run, *self.args, **self.kwargs)
+            self.greenlet.name = f'{self.__class__.__name__}|{self.greenlet.name}'
+        self.greenlet.start()
+
+    def run(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/utils/runnable.py]: Yes, I'd do that, just made it without `_` to tell its part of the public interface (or at least, the interface children must care), but can make it private too.

File: raiden/utils/runnable.py [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+        """ Reimplements in children to busy wait here
+
+        This busy wait should be finished gracefully after stop(),
+        or be killed and re-raise on subtasks exception """
+        raise NotImplementedError
+
+    def stop(self):
+        """ Synchronous stop, gracefully tells run() to exit
+
+        May wait subtasks to finish.
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/utils/runnable.py]: Ok, it should be made synchronous

File: raiden/utils/runnable.py [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+        Stop-time exceptions may be raised, run exceptions should not (accessible via get())
+        """
+        raise NotImplementedError
+
+    def on_error(self, subtask: Greenlet):
+        """ Default callback for substasks link_exception
+
+        Default callback re-raises the exception inside run() """
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/utils/runnable.py]: This, but used from children. It's _default_ in the sense that children may use it as a standard callback for its substasks' `link_exception`. There's no default implementation, but there's a default expected implementation (block inside `run` and provide a point for `kill` to re-raise its exception).

File: raiden/raiden_service.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/raiden_service.py]: Imagine the life of a Runnable: `start` -> `run` -> `stop`
- `start` is synchorous, and may raise if something **during** startup didn't go as expected. Think on a port binding when the port is already in use, and it should prevent the task from being actually started (as some pre-start condition failed)
- `runtime` exceptions are raised _during_ "normal" operation, inside `run` blocking call, either by itself (in case its a loop) or by any subtask (through `kill` or `get`)
- `stop` should also be synchronous, and it should fail on **cleanup** exceptions (imagine a `commit` on a DB which failed). No need to re-raise run-time exceptions

File: raiden/ui/cli.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/ui/cli.py]: It's a way for `gevent` to tell greenlets they were killed, and aren't intended to be re-raised (even because `get` doesn't re-raise them). It's a way for `gevent` to communicate with code running inside a task, and is handled differently from _proper_ exceptions inside it.

File: raiden/raiden_service.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (hackaugusto) [reply raiden/raiden_service.py]: The call to `stop` then `get` still don't make sense.

If you want to call `get` because there could have been a "runtime" exception, then the task is dead already and calling `stop` doesn't make sense.

[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/raiden_service.py]: Runtime exceptions **may** be raised **because** of `stop`. It shouldn't be common, but can happen that `stop` tells the greenlet to exit gracefully, but then an exception is raised inside it. The interface still holds.

File: raiden/utils/runnable.py [hunk 31]
```diff
@@ -0,0 +1,71 @@
+from gevent import Greenlet
+from typing import Sequence
+
+
+class Runnable:
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/utils/runnable.py]: I did it because I plan to move to true `Greenlet`s in the future. It requires refactoring and organizing of `cli.py`, `app.py` and `utils/cli.py`, then `App` may **always** be able to restart by being fully capable of reinstantiating all Runnables/Greenlets everywhere (including tests custom instantiations through custom `config`, smoketests, echo node, etc), and doing the required checks. `Runnable` is intended to be an almost drop-in replacement for `Greenlet` so we should be able to easily do the switch once we do [these refactors](https://github.com/raiden-network/raiden/pull/2208/files#r214356369)

File: raiden/utils/runnable.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (hackaugusto) [reply raiden/utils/runnable.py]: But this value is a tuple

https://docs.python.org/3/reference/compound_stmts.html#function-definitions

>  If the form “*identifier” is present, it is initialized to a tuple receiving any excess positional parameters, defaulting to the empty tuple.

[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden/utils/runnable.py]: Uh.. not sure I was thinking it was a `list`, ok, I'll fix, thank you

File: raiden/network/rpc/client.py (hunk 6 above)
Comment (hackaugusto) [reply raiden/network/rpc/client.py]: Okay, could you please remove the exception `RaidenShuttingDown`? This only exists because we were not relying on `GreenletExit`, and it is only raised from the web3 middleware.

**Comment by LefterisJP** [comment]
This also fixed #1149 ?

**Comment by andrevmatos** [comment]
@LefterisJP yes, most of it. `cli.py`/`app.py` still need reorganization, but this removes all the stop timeouts I could find.

File: raiden/network/transport/matrix.py [hunk 32]
```diff
@@ -514,15 +558,15 @@ def _handle_message(self, room, event) -> bool:
                     "Can't parse message binary data",
                     message_data=data,
                     peer_address=pex(peer_address),
-                    exception=ex,
+                    _exc=ex,
```
Comment (hackaugusto) [inline raiden/network/transport/matrix.py]: why was this changed? what is the point of this?

File: raiden/network/transport/matrix.py [hunk 33]
```diff
@@ -711,15 +755,15 @@ def _send_raw(self, receiver_address, data):
         room = self._get_room_for_address(receiver_address)
         if not room:
             return
-        self.log.debug('SEND', room=room, data=data)
+        self.log.debug('SEND', receiver=pex(receiver_address), room=room, data=data)
```
Comment (hackaugusto) [inline raiden/network/transport/matrix.py]: btw, if you can, adding the current node address is quite useful for debugging and we have it done one a good number of logging calls. if you don't mind,next time please add the equivalent of `node=pex(node_address)`

File: raiden/network/transport/matrix.py (hunk 32 above)
Comment (andrevmatos) [reply raiden/network/transport/matrix.py]: `exception` is used internally by `structlog`. This threw a lot of warnings in the unit tests @LefterisJP created as part of #2300 . This fixes it by using a non-conflicting parameter name.

File: raiden/network/transport/matrix.py (hunk 33 above)
Comment (andrevmatos) [reply raiden/network/transport/matrix.py]: This was avoided in `matrix` explicitly because `self.log` is bound to include `current_user=<user_id>`, which is always in the format `@<hex_addr>:<server_hostname>`, which _replaces_ `pex(node_address)` with more information, so there wasn't the need to spam the log entry with redundant information

File: raiden/network/transport/matrix.py (hunk 32 above)
Comment (hackaugusto) [reply raiden/network/transport/matrix.py]: This seems the wrong fix IMO, either use `log.exception` or [`log.<level>(exc_info=exc)`](http://www.structlog.org/en/stable/api.html#structlog.processors.format_exc_info)

Comment (andrevmatos) [reply raiden/network/transport/matrix.py]: That will include the whole stacktrace of the exception, which isn't required here. This approach only logs the `repr` of the exception.

File: raiden/network/transport/matrix.py (hunk 33 above)
Comment (hackaugusto) [reply raiden/network/transport/matrix.py]: I don't agree, the **only** reason to use structlog is to have parseable logs, which can then be added to a database and queried. If the logs don't have consistent fields to query the data is less useful.

Just be consistent with the rest of the code base. 13 letters in a log line that has the whole message body is hardly "spamming" anything.

Comment (andrevmatos) [reply raiden/network/transport/matrix.py]: The fields **are** consistent IMO, all of them include the node address or part of it. Matrix isn't supposed to comply with **conventions** from the Service, it is (or was made to be) a standalone class. If you want to require a given field to be on every entry, you should consider making a PR to `log.bind` the node address in `raiden_service.py`, and import/require/inject/whatever that on every child logger.
Anyway, I don't have hard feelings on this one, agree with you that 13 characters is negligible, so I'll bind and include it here, but above suggestion holds.
